강의 테이블 수정 완료

mingmingmon opened the pull request:
## Summary

> Lecture 테이블의 칼럼을 수정하였습니다.

강의 정보를 경기대학교 쿠티스에서 스크래핑 하던 중, time 서비스의 강의 정보 조회 기능을 개발할 때 필터링용으로 사용하면 좋을 것 같은 칼럼을 추가했습니다.

1. 강의의 Type을 Enum으로 관리했습니다. 아래 사진과 같이 쿠티스에서 총 5개의 분류로 강의 정보를 제공하고 있습니다. 이처럼 이미 경기대학교 학생들이 많이 접한 분류로 필터링 기능이 time 서비스에 제공된다면 UX적으로 좋을 것 같았습니다. 
Enum을 사용한 이유는, 경기대학교 쿠티스에서 학과 개편을 크게 하지 않는 이상 아래와 같은 분류는 오랫동안 고정되어있을 것이라 생각했습니다.
<img width="608" alt="image" src="https://github.com/user-attachments/assets/dae930f0-5f11-4ac0-a73a-8098f224e30d">


2. groupName이라는 칼럼을 추가했습니다. 기존에는 사용자가 major 칼럼을 활용해 본인의 학과를 검색할 것이라 생각했습니다. 그러나 Type이 전공이 아닌 교양인 경우에는 major 칼럼에 해당되는 데이터가 없었습니다. 사진처럼 드롭다운 박스에서 제공하는 내용이 Type이 교양인 경우 유의미했기에 이를 저장하고자 groupName이라는 칼럼을 만들었습니다.
<img width="595" alt="image" src="https://github.com/user-attachments/assets/bc740699-e09d-4de4-9d01-53d3a5933af7">

Type이 전공인 경우 groupName의 정보는 아래와 같이 선택메뉴의 text로 저장해두었습니다.
<img width="581" alt="image" src="https://github.com/user-attachments/assets/e013661b-b2e8-4560-94b3-958a3980a27b">

3. isExceeded 칼럼이 추가되었습니다. 수강초과된 과목인지, 아닌지 판단할 수 있는 칼럼입니다. 수강신청 시즌과 정정시즌이 되면 수강초과가 안 된 과목을 찾아보는 경우가 많습니다. 해당 칼럼으로 필터링 할 수 있는 API를 만들면 필요한 정보를 사용자에게 줄 수 있을 것 같아서 추가했습니다.

 
## Tasks

- 강의 테이블 수정

SongJaeHoonn commented:
과목 정보의 정렬이나 검색, 시간표 작성, 교양 검색이나 전공별 검색들에 필요한 요소들을 직접 찾아보며 확인해봤는데, 빠진 부분 없이 꼼꼼하게 잘 되어 있는 것 같습니다! 정말 고생하셨습니다. 

과목의 대분류인 5개(교양, 전공...)를 enum으로 분류해두신 것과, isExceeded로 초과 정보를 저장하고자 하신 점이 인상깊습니다.

그런데, 전공 과목의 경우에는 이미 campus 컬럼과 major 컬럼이 존재해서 단순히 campus-major의 문자열 조합이라 groupName이 필요 없을 것 같은데 스크래핑할 때에 해당 내용만을 지워내기엔 불가능한 작업인가요?

mingmingmon commented:
> 그런데, 전공 과목의 경우에는 이미 campus 컬럼과 major 컬럼이 존재해서 단순히 campus-major의 문자열 조합이라 groupName이 필요 없을 것 같은데 스크래핑할 때에 해당 내용만을 지워내기엔 불가능한 작업인가요? 아니면 해당 내용이 DB에 추가된다고 해도 그렇게 신경써도 될 부분이 아닌건지 궁금합니다.

``` 
    group_name_input = soup.find('input', {'name': 'hakgwa_name'})
    group_name = group_name_input['value'] if group_name_input['value'] else previous_group_name

    campus = group_name.split('-')[0]
    major = group_name.split('-')[1]
```

위의 코드는 제가 전공과목 스크래핑할 때 정보를 추출한 코드 일부분입니다. campus와 major가 확정이 되고 campus-major를 합쳐서 group_name을 만드는 것이 아니라 오히려 반대 과정으로 데이터를 뽑아내고 있습니다!
경기대학교  쿠티스에서 강의 정보를 조회할 때 폴더를 열어서 전공을 선택하게 되는데, 이때 얻을 수 있는 정보가 "수원주간-컴퓨터공학전공" 과 같은 문자열이었습니다. 따라서 이를 group_name으로 두고 "-"를 기준으로 파싱해서 campus와 major를 뽑아낸 것입니다. campus와 major는 분리되어서 저장이 되어야 사용자의 필터링 요구사항에 맞게 쿼리문을 짤 수 있어서 group_name에서 분리했습니다!

P.S. 전공을 선택하고 조회하는 경우 나오는 table을 보시면 campus와 major 정보가 없기에 group_name에서 추출할 수 밖에 없었습니다!
![image](https://github.com/user-attachments/assets/b577452f-974b-4559-900d-1c768be999f1)

limehee commented:
UI/UX가 어떻게 만들어질지는 모르겠으나, 단편적으로 봤을 때 `campus`와 `major`를 따로 추출한 점은 인상 깊습니다. 인덱싱을 통해 효과적인 필터링이 가능할 것으로 보이네요.

`groupName`의 경우 `campus`, `major`와 데이터가 중복되기 때문에 남겨두는건 불필요하다고 생각합니다.

mingmingmon commented:
@SongJaeHoonn @limehee 
group_name 칼럼의 필요성에 대해서 의견 주셔서 감사합니다. group_name 칼럼을 삭제하는 방향으로 생각해보았지만 그럼에도 "교양"쪽에서 여전히 group_name 칼럼의 내용은 중요한 정보들이 존재합니다.

1. "전공"에서는 group_name이 college_major 조합이라 필요없는 칼럼임이 자명합니다. 
2. "교양"에서는 major 칼럼이 null이고 (전공에 상관 없이 모든 학생이 들을 수 있는 강의들 == 교양) group_name은 또 다른 분류 정보를 담고 있습니다. 이는 강의명인 name과는 또 다른 정보들입니다.
![image](https://github.com/user-attachments/assets/fbab69be-45bb-4365-a2ca-2341627aefcd)

3. 특히나 저번 학기부터 시행되는 것으로 알고 있는 경기대의 필수교양 "진성애" 과목을 학기별로 번갈아가며 특정 단과대들 학생들만 들을 수 있도록 하는 정책 때문에 아래 사진처럼 group_name에 해당하는 칼럼에 수강가능 단과대 정보가 들어있습니다. 
![image](https://github.com/user-attachments/assets/5fe625ee-69da-4b62-88b2-9f2bac4e38e0)


따라서 아예 group_name 칼럼을 삭제시키기에는 교양쪽에서 필요한 정보들이 손실됩니다. 이러한 상황을 바탕으로 "전공"과 "교양" 모두 group_name 칼럼에 대응되는 정보를 저장하고자 합니다.

위의 구체적인 상황을 고려해주시고, 피드백 부탁드립니다!

limehee commented:
교과목에 상관없이 `group_name`의 정보가 `campus`, `major`로 분류 가능한 줄 알았는데, 상황에 따라 그렇지 않은 경우도 있군요.
말씀해주신 것처럼 교양 교과목에서 `group_name`의 존재가 필수적이라면 남겨두는게 더 좋다는 판단입니다.
일관성 있게 전공에서도 `group_name` 정보를 유지하는 편도 나쁘지 않을 것 같네요.
